Let sitemap event listeners modify the sitemap data

mlocati opened the pull request:
Let's allow sitemap event listeners to modify the sitemap elements by simply using event arguments.

KorvinSzanto commented on concrete/src/Page/Sitemap/SitemapWriter.php:
```diff
@@ -292,15 +292,20 @@ public function generate(callable $pulse = null)
                 if ($dispatchElementReady) {
-                    $this->director->dispatch(static::EVENTNAME_ELEMENTREADY, new GenericEvent(['sitemapPage' => $element]));
+                    $event = new GenericEvent(['sitemapPage' => $element], ['sitemapPage' => $element]);
+                    $this->director->dispatch(static::EVENTNAME_ELEMENTREADY, new GenericEvent(['sitemapPage' => $element], ['sitemapPage' => $element]));
```
This needs to have `$event` passed in right?

KorvinSzanto commented on concrete/src/Page/Sitemap/SitemapWriter.php:
```diff
@@ -292,15 +292,20 @@ public function generate(callable $pulse = null)
                         $xmlNode = $element->toXmlElement($xmlDocument);
                         if ($dispatchPageReadyDeprecated && $element instanceof SitemapPage) {
-                            $this->director->dispatch(static::EVENTNAME_PAGEREADY_DEPRECATED, new GenericEvent(['page' => $element->getPage(), 'xmlNode' => $xmlNode]));
+                            $eventArgs = ['page' => $element->getPage(), 'xmlNode' => $xmlNode];
```
It's probably worth not keeping a copy of this in memory, so either unsetting it or passing the array in as a parameter would be good

mlocati commented:
I updated this pull request by adding sitemap-specific events, so that we have a better control on the arguments.
And yes, I added `unset` calls to save some memory (even if PHP should free a variable when its value gets overwritten).

KorvinSzanto commented on concrete/src/Page/Sitemap/Event/ElementReadyEvent.php:
```diff
@@ -0,0 +1,71 @@
+                throw new InvalidArgumentException(t("Sitemap headers can't be omitted."));
+            }
+            if ($this->element instanceof SitemapFooter) {
+                throw new InvalidArgumentException(t("Sitemap footers can't be omitted."));
```
These are added undocumented exceptions, should we add a `@throws` here and on `SitemapWriter::generate`?

KorvinSzanto commented:
This looks good to me, just had one comment

mlocati commented:
> This looks good to me, just had one comment

Well, this PR clashes with #7128 (just merged), but IMHO this #7129 is a bit more complete. @a3020 thoughts?

a3020 commented:
@mlocati Yes it's more complete, I just thought of adding one event at the time, to make it easier to merge.

a3020 commented on concrete/src/Page/Sitemap/Event/DeprecatedPageReadyEvent.php:
```diff
@@ -0,0 +1,76 @@
+use SimpleXMLElement;
+use Symfony\Component\EventDispatcher\GenericEvent;
+
+class DeprecatedPageReadyEvent extends GenericEvent
```
If it's deprecated, why are we adding a new event class for it?

a3020 commented on concrete/src/Page/Sitemap/Event/DeprecatedPageReadyEvent.php:
```diff
@@ -0,0 +1,76 @@
+    /**
+     * {@inheritdoc}
+     *
+     * @see \Symfony\Component\EventDispatcher\GenericEvent::getSubject()
```
Should we add here that this is for backward compatibility?

a3020 commented on concrete/src/Page/Sitemap/SitemapWriter.php:
```diff
@@ -315,7 +326,10 @@ public function generate(callable $pulse = null)
             if ($mode === static::MODE_HIGHMEMORY) {
                 if ($dispatchXmlReady) {
-                    $this->director->dispatch(static::EVENTNAME_XMLREADY, new GenericEvent(['xmlDoc' => $xmlDocument]));
+                    $event = new Event\XmlReadyEvent($xmlDocument);
```
I think Korvin wants these imported at the top? ;)

aembler commented:
Thanks @mlocati – since everyone agrees this is more complete can you update this pull request to

a. Remove or repurpose the code added in #7128 
b. Fix the error introduced by #7128 (probably will be fixed by refactoring/removing the code.) 
c. Fix the merge conflicts that happened from us merging that pull request?

Then I'll take a look at this and merge.

mlocati commented:
@aembler I reverted #7128 and cherry-picked the commits of this PR: it should be ok now.

a3020 commented:
@mlocati thanks for your clarification. So what exactly means the term 'deprecated' in this context?

From what I see, there's not really an alternative to alter the XML in another way. Do you agree? Isn't it just up to the user to decide to use it or not?

mlocati commented:
> @mlocati thanks for your clarification. So what exactly means the term 'deprecated' in this context?

It's deprecated because you have a better way (much less memory hungry) to customize the sitemap nodes.

> From what I see, there's not really an alternative to alter the XML in another way. Do you agree? Isn't it just up to the user to decide to use it or not?

You can hook the `on_sitemap_xml_element` event, and use the `setElement` method of the `ElementReadyEvent` instance, by specifying a custom `SitemapElement` instance.

a3020 commented:
OK, I tested this PR and was able to modify a sitemap element / node. Thanks @mlocati 

PS. This is the script I used to test it:
```php
$app = Application::getFacadeApplication();
$app['director']->addListener('on_sitemap_xml_element', function ($event) {
    /** @var \Concrete\Core\Page\Sitemap\Event\ElementReadyEvent $event */
    $element = $event->getElement();

    // Don't try to modify a header or footer element.
    if (!$element instanceof \Concrete\Core\Page\Sitemap\Element\SitemapPage) {
        return $event;
    }

    /** @var \Concrete\Core\Page\Sitemap\Element\SitemapPage $element */
    $element->setChangeFrequency('just a test');

    return $event;
});
```

mlocati commented:
Thanks for checking, A3!
